src/windows/list.cpp

```diff
@@ -0,0 +1,687 @@
+  "SCSI", "SD", "PCISTOR",
+  "RTSOR", "JMCR", "JMCF", "RIMMPTSK", "RIMSPTSK", "RIXDPTSK",
+  "TI21SONY", "ESD7SK", "ESM7SK", "O2MD", "O2SD", "VIACR"
+};
+
+// List of known virtual disk hardware IDs
+const std::vector<std::string> VHD_HARDWARE_IDS {
+  "Arsenal_________Virtual_",
+  "KernSafeVirtual_________",
+  "Msft____Virtual_Disk____",
+  "VMware__VMware_Virtual_S"
+};
+
+char* WCharToUtf8(const wchar_t* wstr) {
+  char *str = NULL;
+  size_t size = WideCharToMultiByte(CP_UTF8, 0, wstr, -1, NULL, 0, NULL, NULL);
+
+  if (size <= 1) {
+    return NULL;
+  }
+
+  if ((str = reinterpret_cast<char*>(calloc(size, 1))) == NULL) {
+    return NULL;
+  }
+
+  size_t utf8Size = WideCharToMultiByte(
+    CP_UTF8, 0, wstr, -1, str, size, NULL, NULL);
+
+  if (utf8Size != size) {
+    free(str);
+    return NULL;
+  }
+
+  return str;
+}
+
+char* GetEnumeratorName(HDEVINFO hDeviceInfo, SP_DEVINFO_DATA deviceInfoData) {
+  DWORD size;
+  DWORD dataType;
+  char buffer[MAX_PATH];
+
+  ZeroMemory(&buffer, sizeof(buffer));
+
+  BOOL hasEnumeratorName = SetupDiGetDeviceRegistryPropertyA(
+    hDeviceInfo, &deviceInfoData, SPDRP_ENUMERATOR_NAME,
+    &dataType, (LPBYTE) buffer, sizeof(buffer), &size);
+
+  return hasEnumeratorName ? buffer : NULL;
+}
+
+std::string GetFriendlyName(HDEVINFO hDeviceInfo,
+  SP_DEVINFO_DATA deviceInfoData) {
+  DWORD size;
+  DWORD dataType;
+  wchar_t wbuffer[MAX_PATH];
+
+  ZeroMemory(&wbuffer, sizeof(wbuffer));
+
+  BOOL hasFriendlyName = SetupDiGetDeviceRegistryPropertyW(
+    hDeviceInfo, &deviceInfoData, SPDRP_FRIENDLYNAME,
+    &dataType, (PBYTE) wbuffer, sizeof(wbuffer), &size);
+
+  return hasFriendlyName ? WCharToUtf8(wbuffer) : std::string("");
+}
+
+bool IsSCSIDevice(std::string enumeratorName) {
+  for (std::string driverName : GENERIC_STORAGE_DRIVERS) {
+    if (enumeratorName == driverName) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool IsUSBDevice(std::string enumeratorName) {
+  for (std::string driverName : USB_STORAGE_DRIVERS) {
+    if (enumeratorName == driverName) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool IsRemovableDevice(HDEVINFO hDeviceInfo, SP_DEVINFO_DATA deviceInfoData) {
+  DWORD size;
+  DWORD dataType;
+  DWORD result = 0;
+
+  BOOL hasRemovalPolicy = SetupDiGetDeviceRegistryProperty(
+    hDeviceInfo, &deviceInfoData, SPDRP_REMOVAL_POLICY,
+    &dataType, (PBYTE) &result, sizeof(result), &size);
+
+  switch (result) {
+    case CM_REMOVAL_POLICY_EXPECT_SURPRISE_REMOVAL:
+    case CM_REMOVAL_POLICY_EXPECT_ORDERLY_REMOVAL:
+      return true;
+    default:
+      return false;
+  }
+
+  return false;
+}
+
+bool IsVirtualHardDrive(HDEVINFO hDeviceInfo, SP_DEVINFO_DATA deviceInfoData) {
+  DWORD size;
+  DWORD dataType;
+  char buffer[MAX_PATH];
+
+  ZeroMemory(&buffer, sizeof(buffer));
+
+  BOOL hasHardwareId = SetupDiGetDeviceRegistryPropertyA(
+    hDeviceInfo, &deviceInfoData, SPDRP_HARDWAREID,
+    &dataType, (LPBYTE) buffer, sizeof(buffer), &size);
```

Comment:
Given `dataType` and `size` is not used (I think this is also the case on similar functions on this file), could we pass `NULL` to `SetupDiGetDeviceRegistryPropertyA`?